EGL: add support for WebGL context attributes

moralrecordings opened the pull request:
WebGL allows limited configuration of the rendering context with a set of boolean [context attributes](https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/getContext). The equivalent in EGL would be some of the attributes passed to [eglChooseConfig](https://www.khronos.org/registry/egl/sdk/docs/man/html/eglChooseConfig.xhtml) before creating the context. Right now the EGL shim has the 'depth', 'stencil' and 'antialias' attributes hardcoded to true by default, which can cause issues (e.g. https://github.com/kripken/emscripten/issues/4730). Patch makes EGL watch for positive requests for EGL_DEPTH_SIZE, EGL_STENCIL_SIZE, EGL_SAMPLE_BUFFERS and EGL_ALPHA_SIZE, and passes them as flags to glutInitDisplayMode, which in turn passes them to HTMLCanvasElement.getContext

Tests run: browser.test_egl*, browser.test_gl*, browser.test_sdl*

kripken commented on src/library_egl.js:
```diff
@@ -27,11 +33,34 @@ var LibraryEGL = {
         EGL.setErrorCode(0x300C /* EGL_BAD_PARAMETER */);
         return 0;
       }
+      for(;;) {
```
For consistency with other code, please do `while (1) {`

kripken commented on src/library_egl.js:
```diff
@@ -136,10 +171,10 @@ var LibraryEGL = {
       return 1;
     case 0x3025: // EGL_DEPTH_SIZE
-      {{{ makeSetValue('value', '0', '24' /* 24 bits for depth buffer. TODO: This is hardcoded, add support for this! */, 'i32') }}};
+      {{{ makeSetValue('value', '0', 'EGL.contextAttributes.depth?24:0' /* 24 bits for depth buffer. */, 'i32') }}};
```
space before and after `?` and `:`

kripken commented:
Overall looks good to me, thanks. @juj knows best though for this stuff.

Could we add a test for this?

juj commented:
Thanks for working on this @moralrecordings! Unfortunately the implementation is not in the right direction. The main problem is that `eglChooseConfig` is supposed to be an immutable/static getter function that can be queried to receive a list of supported configurations that matches a given filter.

However this pull request implements `eglChooseConfig` to behave as if it were some kind of setter function that specified the kind of mode to enable. That is not how the function should work.

To be correct, calling `eglChooseConfig` should not modify any internal state, but should be read-only (i.e. as if it was `const` in C++). The documentation for eglChooseConfig is [here](https://www.khronos.org/registry/egl/sdk/docs/man/html/eglChooseConfig.xhtml), and full EGL spec [here](https://www.khronos.org/registry/egl/specs/eglspec.1.4.20110406.pdf).

moralrecordings commented:
Fair enough, it isn't the best fit for the spec. Would a better approach be to generate a static framebuffer configuration for each combination of options (16 all up), return the filtered choices with eglChooseConfig, then rework eglCreateContext, eglGetConfigAttrib etc. to read from that static set instead? If so, is it alright to drop the magic 62002 number and instead return config IDs from 0-11?

moralrecordings commented:
Also thanks for pulling me up there, I'm more familiar with SDL2's EGL wrapper so like a dunce I went ahead and did it the way SDL2 does without stopping to think.

juj commented:
Yeah, a good way is to generate a list of all the combinations and use the list index of a particular config as its ID. (and add +1 to keep id=0 as invalid) This will make the magic ID 62002 obsolete.

juj commented on src/library_egl.js:
```diff
@@ -15,6 +15,28 @@ var LibraryEGL = {
+    currentEGLConfig: 0,
+    // List of different EGLConfig configurations in spec order
+    configAttribs: [
+        {alpha: 8,  samples: 0, buffers: 0,  depth: 0,   stencil: 0},
```
Perhaps this would be better to be a nested set of for loops, that way in the future the code size won't explode as new configs are added. E.g.

    for(var alpha of [0, 8]) {
      for(var samples of [0, 4]) {
        for( ... ) {
          configs.push({alpha: alpha, samples: samples, ... });
        }
      }
    }

Skimming through the EGL spec, the caps `EGL_ALPHA_SIZE`, `EGL_CONFIG_CAVEAT`, `EGL_DEPTH_SIZE`, `EGL_STENCIL_SIZE`, `EGL_SAMPLE_BUFFERS`, `EGL_SAMPLES`, `EGL_MIN_SWAP_INTERVAL`, `EGL_MAX_SWAP_INTERVAL` and `EGL_TRANSPARENT_TYPE` are items that have a meaning for WebGL as well, and even if we don't add all of them right now, those could be added later, so good to have a compact representation that is easily extended.

juj commented on src/library_egl.js:
```diff
@@ -27,16 +49,69 @@ var LibraryEGL = {
-        {{{ makeSetValue('config', '0', '62002' /* Magic ID for the only EGLConfig supported by Emscripten */, 'i32') }}}; 
+      var configCount = 0;
+      for (var i=0; i<EGL.configAttribs.length; i++) {
+        if ((EGL.configAttribs[i].alpha >= alpha) &&
```
No need for the inner parentheses, i.e. `if (x >= y &&` is sufficient.

juj commented on src/library_egl.js:
```diff
@@ -27,16 +49,69 @@ var LibraryEGL = {
-      if (config && config_size > 0) {
-        {{{ makeSetValue('config', '0', '62002' /* Magic ID for the only EGLConfig supported by Emscripten */, 'i32') }}}; 
+      var configCount = 0;
+      for (var i=0; i<EGL.configAttribs.length; i++) {
```
Space after `=` and `<`, i.e. `for (var i = 0; i < EGL.configAttribs.length; i++) {`.

juj commented on src/library_egl.js:
```diff
@@ -27,16 +49,69 @@ var LibraryEGL = {
+            (EGL.configAttribs[i].buffers >= buffers) &&
+            (EGL.configAttribs[i].depth >= depth) &&
+            (EGL.configAttribs[i].stencil >= stencil)) {
+          if (config) {
```
Indentation dropped here.

juj commented on src/library_egl.js:
```diff
@@ -27,16 +49,69 @@ var LibraryEGL = {
+          var value = {{{ makeGetValue('attribList', '4', 'i32') }}};
+          switch (param) {
+            case 0x3021: // EGL_ALPHA_SIZE
+              alpha = value;
```
It would be good to recognize `EGL_DONT_CARE` here. With this code that would get treated as an invalid value that no config matches, returning 0 configurations if any `EGL_DONT_CARE`s are specified.

moralrecordings replied:
Technically EGL_DONT_CARE would still work, as it is defined as -1 and the comparisons are all of the form (configAttr >= value) . But I agree that it's worth singling out.

juj commented on src/library_egl.js:
```diff
@@ -110,7 +186,7 @@ var LibraryEGL = {
       return 0;
     }
-    if (config != 62002 /* Magic ID for the only EGLConfig supported by Emscripten */) {
+    if ((config <= 0) || (config > EGL.configAttribs.length)) { /* Check EGLConfig ID is within acceptable range */
```
It would be good to do `var config_index = config - 1;` early on here so that the ID to index conversion arithmetic is done exactly once up front, and all other code locations can then index directly and not have to repeat the subtract by one. `config_index < 0 || config_index >= EGL.configAttribs.length` would then look a bit more idiomatic than fusing the subtract one arithmetic into the comparison operators here. Also could store `EGL.configAttribs[config_index]` to a variable to avoid repeating that below.

juj commented on src/library_egl.js:
```diff
@@ -289,7 +370,16 @@ var LibraryEGL = {
+      (EGL.configAttribs[config-1].buffers && 0x0080) /* GLUT_MULTISAMPLE */
+    )
+    
+    _glutInitDisplayMode(displayMode);
```
The reason why this code path calls to GLUT to initialize the GL context is historical since back then we didn't have a good other API to create a context, and in the long run we certainly don't want to do this anymore.

Since then we've added a "native" WebGL context creation function `GL.createContext`, which is available here: https://github.com/kripken/emscripten/blob/master/src/library_gl.js#L405. See here for an example of how that is called: https://github.com/kripken/emscripten/blob/master/src/library_html5.js#L1726. If you're interested to doing this change, feel free to, although certainly can leave that separate for later as well.

juj commented:
Very nice, this is looking like the right direction now!

scottp-dpaw commented on src/library_egl.js:
```diff
@@ -136,17 +212,17 @@ var LibraryEGL = {
       return 1;
     case 0x3025: // EGL_DEPTH_SIZE
-      {{{ makeSetValue('value', '0', '24' /* 24 bits for depth buffer. TODO: This is hardcoded, add support for this! */, 'i32') }}};
+      {{{ makeSetValue('value', '0', 'EGL.configAttribs[config-1].depth' /* 16 bits for depth buffer. */, 'i32') }}};
```
https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.2.1 - you're right in that desktop browsers do generally go for 24 bits. The spec only promises "at least 16 bits" as that's lower limit GLESv2 guarantees to provide, which definitely is the case on mobile chips. (leaving aside that Emscripten apps run like a sack of wet hammers on mobile)

moralrecordings commented on src/library_egl.js:
```diff
@@ -27,16 +49,69 @@ var LibraryEGL = {
+      if (attribList) {
+        while (1) {
+          var param = {{{ makeGetValue('attribList', '0', 'i32') }}};
+          if (param == 0x3038) { // EGL_NONE
```
Makes sense. While we're at it, would it be a good idea to move all of the EGL constants that are sprinkled about as magic numbers into the EGL object, e.g. EGL.DEPTH_SIZE, EGL.STENCIL_SIZE? Wouldn't have to be all 170 obviously, just the subset that are used.

moralrecordings replied:
I added a sanity check for the attribute ID parameter. This should capture the vast majority of bad eglChooseConfig calls; I toyed with checking the values as well, but the EGL spec is extremely vague when it comes to validation requirements.

moralrecordings commented:
Sorry for leaving this so long, life comes at you fast. I think I've addressed all of the review points now, let me know if there's anything left that I've missed.

kripken commented:
ping @juj

juj commented:
Looks good, sorry for being unavailable to review this for a long time. If this passes the CI after rebase, lgtm to land!

sbc100 commented:
The deletion of the `incoming` branch triggered the automatic closing of this branch.  My apologies.  This was an unexpected side effect.

Re-opening and re-targeting to master.
